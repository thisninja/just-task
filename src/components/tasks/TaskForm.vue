<template>
  <div class="card-body">
    <div class="card-text">
      <span>
        {{ currentTitle }}
      </span>
      <md-datepicker
        md-immediately
        v-model="selectedDate"
        :md-disabled-dates="disabledDates"
        >
        <label>
          {{ DUE_DATE_TEXT }}
        </label>
      </md-datepicker>
      <md-field>
        <label>
          {{ TASK_DESCRIPTION_TEXT }}
        </label>
        <md-input
          v-model="text"
          type="text"
          >
        </md-input>
      </md-field>
      <md-button
        class="fn-save"
        :disabled="disabled"
        @click="targetHandler"
        >
        {{ SAVE_BTN_TEXT }}
      </md-button>
      <md-button
        class="fn-cancel"
        @click="onCancel"
        >
        {{ CANCEL_BTN_TEXT }}
      </md-button>
    </div>
  </div>
</template>

<script>
import {
  mapActions,
} from 'vuex';

import {
  DUE_DATE_TEXT,
  ADD_TASK_TEXT,
  EDIT_TASK_TEXT,
  TASK_DESCRIPTION_TEXT,
  SAVE_BTN_TEXT,
  CANCEL_BTN_TEXT,
} from './constants';

export default {
  props: ['id', 'currentText', 'dueDate'],
  name: 'TaskForm',
  data() {
    return {
      text: this.currentText,
      selectedDate: this.dueDate,
      disabledDates: date => {
        return date.getTime() < Date.now()
      },
      DUE_DATE_TEXT,
      TASK_DESCRIPTION_TEXT,
      SAVE_BTN_TEXT,
      CANCEL_BTN_TEXT,
    };
  },
  computed: {
    currentTitle() {
      return this.id
        ? EDIT_TASK_TEXT
        : ADD_TASK_TEXT;
    },
    disabled() {
      return (!this.text || !this.selectedDate);
    },
  },
  methods: {
    ...mapActions([
      'addNewTask',
      'updateTask',
    ]),
    targetHandler() {
      return this.id
        ? this.editTask()
        : this.addTask();
    },
    addTask(){
      this.addNewTask({
        text: this.text,
        dueDate: this.selectedDate
      });

      this.$emit('update');
    },
    editTask() {
      this.updateTask({
        _id: this.id,
        text: this.text,
        dueDate: new Date(this.selectedDate).getTime(),
      });

      this.$emit('update');
    },
    onCancel() {
      this.$emit('update');
    }
  }
};
</script>
