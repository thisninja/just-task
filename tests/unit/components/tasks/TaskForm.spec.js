import Vuex from 'vuex';
import VueMaterial from 'vue-material';
import { mount, createLocalVue, shallowMount } from '@vue/test-utils';
import TaskForm from '@/components/tasks/TaskForm.vue';
import dataMock from './__mocks__/dataMock';
import {
  ADD_TASK_TEXT,
  EDIT_TASK_TEXT,
} from '@/components/tasks/constants';

const localVue = createLocalVue();

localVue.use(VueMaterial);
localVue.use(Vuex);

describe('TaskForm.vue', () => {
  let wrapper;
  let store;
  let actions;
  let mutations;
  let getters;

  beforeEach(() => {
    getters = getters || {
      tasks: () => [],
    };

    actions = actions || {
      addNewTask: jest.fn(),
      updateTask: jest.fn(),
    };

    store = new Vuex.Store({
      getters,
      actions,
    });
  });

  describe('Add TaskForm', () => {
    beforeAll(() => {
      wrapper = shallowMount(TaskForm, {
        store,
        localVue,
        propsData: {
          currentText: dataMock.tasks[0].text,
          dueDate: dataMock.tasks[0].dueDate,
        },
      });
    });

    it('should match snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('Edit TaskForm', () => {
    beforeAll(() => {
      wrapper = shallowMount(TaskForm, {
        store,
        localVue,
        propsData: {
          id: dataMock.tasks[1]._id,
          currentText: dataMock.tasks[1].text,
          dueDate: dataMock.tasks[1].dueDate,
        },
      });
    });

    it('should match snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('computed', () => {
    it('currentTitle', () => {
      expect(wrapper.vm.currentTitle).toEqual(EDIT_TASK_TEXT);
    });

    it('disabled', () => {
      expect(wrapper.vm.disabled).toEqual(false);

      wrapper.vm.text = '';
      wrapper.vm.selectedDate = null;

      expect(wrapper.vm.disabled).toEqual(true);
    });
  });

  describe('methods', () => {
    it('targetHandler', () => {
      wrapper.vm.editTask = jest.fn();
      wrapper.vm.targetHandler();

      expect(wrapper.vm.editTask).toHaveBeenCalled();
    });

    it('addTask', () => {
      wrapper.vm.addTask();
      expect(wrapper.emitted().update).toBeTruthy();
    });

    it('editTask', () => {
      wrapper.vm.editTask();
      expect(wrapper.emitted().update).toBeTruthy();
    });

    it('onCancel', () => {
      wrapper.vm.onCancel();
      expect(wrapper.emitted().update).toBeTruthy();
    });
  });
});
