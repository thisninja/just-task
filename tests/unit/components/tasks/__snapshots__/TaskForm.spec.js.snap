// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm.vue Add TaskForm should match snapshot 1`] = `
<div class="card-body">
  <div class="card-text"><span>
      Add task
    </span>
    <md-datepicker-stub value="1572123600000" mddisableddates="function disabledDates(date) {
        return date.getTime() < Date.now();
      }" mdopenonfocus="true" mdoverridenative="true" mdimmediately="true" mdmodeltype="function Date() { [native code] }" mddebounce="1000"><label>
        Due date
      </label></md-datepicker-stub>
    <md-field-stub mdcounter="true" mdtogglepassword="true"><label>
        Enter your task description
      </label>
      <md-input-stub value="Do someting 1st time" id="md-input-4q9d28au" type="text"></md-input-stub>
    </md-field-stub>
    <md-button-stub mdripple="true" type="button" class="fn-save">
      Save
    </md-button-stub>
    <md-button-stub mdripple="true" type="button" class="fn-cancel">
      Cancel
    </md-button-stub>
  </div>
</div>
`;

exports[`TaskForm.vue Edit TaskForm should match snapshot 1`] = `
<div class="card-body">
  <div class="card-text"><span>
      Edit task
    </span>
    <md-datepicker-stub value="1572300000000" mddisableddates="function disabledDates(date) {
        return date.getTime() < Date.now();
      }" mdopenonfocus="true" mdoverridenative="true" mdimmediately="true" mdmodeltype="function Date() { [native code] }" mddebounce="1000"><label>
        Due date
      </label></md-datepicker-stub>
    <md-field-stub mdcounter="true" mdtogglepassword="true"><label>
        Enter your task description
      </label>
      <md-input-stub value="Do someting 2nd time" id="md-input-ecmezbpbq" type="text"></md-input-stub>
    </md-field-stub>
    <md-button-stub mdripple="true" type="button" class="fn-save">
      Save
    </md-button-stub>
    <md-button-stub mdripple="true" type="button" class="fn-cancel">
      Cancel
    </md-button-stub>
  </div>
</div>
`;
